lib/validator/v3/validator.go

```diff
@@ -695,6 +696,32 @@ func validateRule(v *validator.Validate, structLevel *validator.StructLevel) {
 		}
 	}
 
+	if rule.ICMP != nil && (rule.Protocol == nil || (*rule.Protocol != numorstring.ProtocolFromString("ICMP") &&
+		*rule.Protocol != numorstring.ProtocolFromString("ICMPv6"))) {
+		structLevel.ReportError(reflect.ValueOf(rule.ICMP), "ICMP", "", reason(protocolIcmpMsg))
+	}
+
+	// Check that the IPVersion of the protocol matches the IPVersion of the rule for
+	// known protocols.
+	if rule.IPVersion != nil {
```

Comment:
I think you should do this validation the other way around for ICMP:
```
if protocol == "ICMP" && ipVersion != 4 {
    // invalid
} else if protocol == "ICMPv6" && ipVersion != 6 {
    // invalid
}
```
(I don't think Felix would work with "ICMP" and no IP version either.)

It's sad that ICMP ended up having different protocol names for v4 and v6 but that's what we've got so if you use the v4 protocol name, your rule needs to specify IPv4 :-(